test/integration/scheduler/queue_test.go

```diff
@@ -344,6 +344,43 @@ func TestCoreResourceEnqueue(t *testing.T) {
 			},
 			wantRequeuedPods: sets.New("pod1"),
 		},
+		{
+			name:         "Pod rejected by the TaintToleration plugin is requeued when the Node's taint is updated",
```

Comment:
Can we also add these two test cases:

1. The rejected pod is requeued after adding a toleration.
2. The rejected pod is **not**  requeued with a `UpdateNodeTaint` event, if the pod's toleration already matched the old node's taint.

For case 2, you'll need to let the pod rejected by another plugin. Maybe you can refer to https://github.com/kubernetes/kubernetes/blob/b2031b3cb46e946ee72eab7bda87b046db138d62/test/integration/scheduler/queue_test.go#L424

Reply:
Thanks for your review.

> The rejected pod is requeued after adding a toleration.

I thought it's unnecessary because it is duplicated in https://github.com/kubernetes/kubernetes/blob/b2031b3cb46e946ee72eab7bda87b046db138d62/test/integration/scheduler/queue_test.go#L331


> The rejected pod is not requeued with a UpdateNodeTaint event, if the pod's toleration already matched the old node's taint.

I tried to create the following case, but in this case the pod is requeued.

```go
		{
			name: "Pod that has the corresponding toleration rejected by the NodeResourceFit plugin isn't requeued when the Node's taint is updated",
			initialNodes: []*v1.Node{
				st.MakeNode().Name("fake-node").Capacity(map[v1.ResourceName]string{v1.ResourceCPU: "2"}).Taints([]v1.Taint{{Key: v1.TaintNodeNotReady, Effect: v1.TaintEffectNoSchedule}}).Obj(),
			},
			pods: []*v1.Pod{
				st.MakePod().Name("pod1").Toleration(v1.TaintNodeNotReady).Req(map[v1.ResourceName]string{v1.ResourceCPU: "4"}).Container("image").Obj(),
			},
			triggerFn: func(testCtx *testutils.TestContext) error {
				if _, err := testCtx.ClientSet.CoreV1().Nodes().Update(testCtx.Ctx, st.MakeNode().Name("fake-node").Capacity(map[v1.ResourceName]string{v1.ResourceCPU: "2"}).Taints([]v1.Taint{{Key: "taint-key", Effect: v1.TaintEffectNoSchedule}}).Obj(), metav1.UpdateOptions{}); err != nil {
					return fmt.Errorf("failed to update the Node: %w", err)
				}
				return nil
			},
			wantRequeuedPods:          sets.Set[string]{},
		},
```

Please let me know if I have misunderstood.

Reply:
> I thought it's unnecessary because it is duplicated 

Ah, yes it is.

Reply:
> I tried to create the following case, but in this case the pod is requeued.

Ah, all right. So it seems that we'll register default QHint for any plugin that doesn't register nodeUpdateTaint event.
https://github.com/kubernetes/kubernetes/blob/36122d5a9bbb87f484149e3eb11bb4208228c36f/pkg/scheduler/scheduler.go#L433-L443

So I guess any plugin will return `Queue` with UpdateTaint and let the pod requeued.

We can forget this case, we may need to add this after `preCheck` is fully removed.

Instead, let's add another simple case for "taint updated, but pod still can't tolerate it"
```
			initialNodes: []*v1.Node{
				st.MakeNode().Name("fake-node").Taints([]v1.Taint{{Key: "foo", Effect: v1.TaintEffectNoSchedule}}).Obj(),
			},
			pods: []*v1.Pod{
				st.MakePod().Name("pod1").Toleration(v1.TaintNodeNotReady).Container("image").Obj(),
			},
			triggerFn: func(testCtx *testutils.TestContext) error {
				if _, err := testCtx.ClientSet.CoreV1().Nodes().Update(testCtx.Ctx, st.MakeNode().Name("fake-node").Taints([]v1.Taint{{Key: "bar", Effect: v1.TaintEffectNoSchedule}}).Obj(), metav1.UpdateOptions{}); err != nil {
					return fmt.Errorf("failed to update the Node: %w", err)
				}
				return nil
			},
```

I'm not sure if we already have this case(I remember we lack many `not requeued` test case)

Reply:
> I thought it's unnecessary because it is duplicated in
> 
> https://github.com/kubernetes/kubernetes/blob/b2031b3cb46e946ee72eab7bda87b046db138d62/test/integration/scheduler/queue_test.go#L331

Could you add a second pod there that won't be requeued after pod1 update?

Reply:
> So I guess any plugin will return Queue with UpdateTaint and let the pod requeued.

Thank you for noticing. I thought your guess was correct because when I removed the if block and ran the case, the pod was not requeued and the test case passed.

> We can forget this case, we may need to add this after preCheck is fully removed.

Okey, for now, I will skip adding this case. 

> Instead, let's add another simple case for "taint updated, but pod still can't tolerate it"

This case was covered by adding pod3 with unmatched toleration to the existing UpdateTaint case.

Reply:
> Could you add a second pod there that won't be requeued after pod1 update?

Thanks for the review. I added it.